File: packages/javascript/test/options.test.ts
import assert from 'node:assert';
import { describe, it } from 'node:test';
import {
	ANALYZE_CHECK_CODES,
	argsToOptions,
	buildUsageText,
	CHECK_CODE,
	DEFAULT_TARGET_LOCALE,
	FILE_FORMAT,
	OPTION_DEFINITIONS,
	resolveOptions
} from '../dist/index.js';

describe('resolveOptions', () => {
	it('fills in the defaults', () => {
		const { options } = resolveOptions();

		assert.strictEqual(options.target, DEFAULT_TARGET_LOCALE);
		assert.strictEqual(options.format, FILE_FORMAT.AUTO);
		assert.strictEqual(options.interpolationPrefix, '{');
		assert.strictEqual(options.enabledChecks.size, ANALYZE_CHECK_CODES.length);
	});

	it('does not treat a missing target as a fault', () => {
		const { issues } = resolveOptions();

		assert.strictEqual(issues.length, 1);
		assert.strictEqual(issues[0].level, 'info');
	});

	it('accepts a comma separated list of check codes', () => {
		const { options } = resolveOptions({ checks: 'NO_KEY, EMPTY_VALUE' });

		assert.deepStrictEqual([...options.enabledChecks], [CHECK_CODE.NO_KEY, CHECK_CODE.EMPTY_VALUE]);
	});

	it('removes the ignored checks from the full set', () => {
		const { options } = resolveOptions({ ignoreChecks: ['DUPLICATE_VALUE'] });

		assert.strictEqual(options.enabledChecks.has(CHECK_CODE.DUPLICATE_VALUE), false);
		assert.strictEqual(options.enabledChecks.size, ANALYZE_CHECK_CODES.length - 1);
	});

	it('reports an unknown check code and keeps going', () => {
		const { options, issues } = resolveOptions({ checks: 'NO_KEY,NOPE' });

		assert.deepStrictEqual([...options.enabledChecks], [CHECK_CODE.NO_KEY]);
		assert.ok(issues.some((issue) => issue.message.includes('NOPE')));
	});

	it('refuses to combine checks with ignoreChecks', () => {
		const { options, issues } = resolveOptions({
			checks: 'NO_KEY',
			ignoreChecks: 'EMPTY_VALUE'
		});

		assert.deepStrictEqual([...options.enabledChecks], [CHECK_CODE.NO_KEY]);
		assert.ok(issues.some((issue) => issue.message.includes('ignoreChecks')));
	});

	it('falls back to auto for an unknown format', () => {
		const { options, issues } = resolveOptions({ format: 'nope' as never });

		assert.strictEqual(options.format, FILE_FORMAT.AUTO);
		assert.ok(issues.some((issue) => issue.message.includes('nope')));
	});
});

describe('argsToOptions', () => {
	it('maps every CLI flag onto its option', () => {
		const options = argsToOptions({
			_: [],
			path: 'locales',
			target: 'ko',
			format: 'folder',
			'ignore-checks': 'NO_KEY',
			'interpolation-prefix': '{{',
			'interpolation-suffix': '}}',
			exclude: 'tmp',
			warn: false,
			debug: true
		});

		assert.deepStrictEqual(options, {
			path: 'locales',
			target: 'ko',
			format: 'folder',
			ignoreChecks: 'NO_KEY',
			interpolationPrefix: '{{',
			interpolationSuffix: '}}',
			exclude: 'tmp',
			warn: false,
			debug: true
		});
	});

	it('reads a bare positional argument as the path', () => {
		assert.strictEqual(argsToOptions({ _: ['locales'] }).path, 'locales');
	});

	it('prefers an explicit path over the positional argument', () => {
		assert.strictEqual(argsToOptions({ _: ['ignored'], path: 'locales' }).path, 'locales');
	});

	it('resolves the CLI form of an option exactly like the API form', () => {
		const fromCli = resolveOptions(
			argsToOptions({ _: [], target: 'ko', 'ignore-checks': 'NO_KEY' })
		);
		const fromApi = resolveOptions({ target: 'ko', ignoreChecks: ['NO_KEY'] });

		assert.deepStrictEqual(fromCli.options, fromApi.options);
	});
});

describe('buildUsageText', () => {
	it('documents every option', () => {
		const usage = buildUsageText('chki18n');

		for (const definition of OPTION_DEFINITIONS) {
			assert.ok(usage.includes(`--${definition.flag}`), `${definition.flag} is undocumented`);
		}
	});
});

File: packages/dart/test/options_test.dart
import 'package:chki18n/chki18n.dart';
import 'package:test/test.dart';

void main() {
  group('resolveOptions', () {
    test('fills in the defaults', () {
      final resolved = resolveOptions();

      expect(resolved.options.target, defaultTargetLocale);
      expect(resolved.options.format, Chki18nFileFormat.auto);
      expect(resolved.options.interpolationPrefix, '{');
      expect(resolved.options.enabledChecks.length, analyzeCheckCodes.length);
    });

    test('does not treat a missing target as a fault', () {
      final resolved = resolveOptions();

      expect(resolved.issues.length, 1);
      expect(resolved.issues.first.level, Chki18nLevel.info);
    });

    test('accepts a comma separated list of check codes', () {
      final resolved = resolveOptions(
        const Chki18nOptions(text: Chki18nTextOptions(checks: 'NO_KEY, EMPTY_VALUE')),
      );

      expect(resolved.options.enabledChecks.toList(), [
        Chki18nCheckCode.noKey,
        Chki18nCheckCode.emptyValue,
      ]);
    });

    test('removes the ignored checks from the full set', () {
      final resolved = resolveOptions(
        const Chki18nOptions(ignoreChecks: [Chki18nCheckCode.duplicateValue]),
      );

      expect(resolved.options.enabledChecks.contains(Chki18nCheckCode.duplicateValue), isFalse);
      expect(resolved.options.enabledChecks.length, analyzeCheckCodes.length - 1);
    });

    test('reports an unknown check code and keeps going', () {
      final resolved = resolveOptions(
        const Chki18nOptions(text: Chki18nTextOptions(checks: 'NO_KEY,NOPE')),
      );

      expect(resolved.options.enabledChecks.toList(), [Chki18nCheckCode.noKey]);
      expect(resolved.issues.any((issue) => issue.message.contains('NOPE')), isTrue);
    });

    test('refuses to combine checks with ignoreChecks', () {
      final resolved = resolveOptions(
        const Chki18nOptions(
          text: Chki18nTextOptions(checks: 'NO_KEY', ignoreChecks: 'EMPTY_VALUE'),
        ),
      );

      expect(resolved.options.enabledChecks.toList(), [Chki18nCheckCode.noKey]);
      expect(resolved.issues.any((issue) => issue.message.contains('ignoreChecks')), isTrue);
    });

    test('falls back to auto for an unknown format', () {
      final resolved = resolveOptions(
        const Chki18nOptions(text: Chki18nTextOptions(format: 'nope')),
      );

      expect(resolved.options.format, Chki18nFileFormat.auto);
      expect(resolved.issues.any((issue) => issue.message.contains('nope')), isTrue);
    });

    test('reads a `CODE=level` pair the way the CLI writes it', () {
      final resolved = resolveOptions(
        const Chki18nOptions(text: Chki18nTextOptions(levels: 'EMPTY_VALUE=error')),
      );

      expect(resolved.options.levels, {Chki18nCheckCode.emptyValue: Chki18nLevel.error});
    });

    test('refuses to re-grade a check that reports how the run went', () {
      final resolved = resolveOptions(
        const Chki18nOptions(levels: {Chki18nCheckCode.invalidFile: Chki18nLevel.info}),
      );

      expect(resolved.options.levels, isNull);
      expect(resolved.issues.any((issue) => issue.message.contains('INVALID_FILE')), isTrue);
    });
  });

  group('optionsFromArgs', () {
    test('maps every CLI flag onto its option', () {
      final options = optionsFromArgs({
        '_': <String>[],
        'path': 'locales',
        'target': 'ko',
        'format': 'folder',
        'ignore-checks': 'NO_KEY',
        'interpolation-prefix': '{{',
        'interpolation-suffix': '}}',
        'exclude': 'tmp',
        'warn': false,
        'debug': true,
      });

      expect(options.path, 'locales');
      expect(options.target, 'ko');
      expect(options.text?.format, 'folder');
      expect(options.text?.ignoreChecks, 'NO_KEY');
      expect(options.interpolationPrefix, '{{');
      expect(options.interpolationSuffix, '}}');
      expect(options.text?.exclude, 'tmp');
      expect(options.warn, isFalse);
      expect(options.debug, isTrue);
    });

    test('reads a bare positional argument as the path', () {
      expect(
        optionsFromArgs({
          '_': ['locales'],
        }).path,
        'locales',
      );
    });

    test('prefers an explicit path over the positional argument', () {
      expect(
        optionsFromArgs({
          '_': ['ignored'],
          'path': 'locales',
        }).path,
        'locales',
      );
    });

    test('resolves the CLI form of an option exactly like the API form', () {
      final fromCli = resolveOptions(
        optionsFromArgs({'_': <String>[], 'target': 'ko', 'ignore-checks': 'NO_KEY'}),
      );
      final fromApi = resolveOptions(
        const Chki18nOptions(target: 'ko', ignoreChecks: [Chki18nCheckCode.noKey]),
      );

      expect(fromCli.options.toJson(), fromApi.options.toJson());
    });
  });

  group('buildUsageText', () {
    test('documents every option', () {
      final usage = buildUsageText('chki18n');

      for (final definition in optionDefinitions) {
        expect(
          usage,
          contains('--${definition.flag}'),
          reason: '${definition.flag} is undocumented',
        );
      }
    });
  });
}
